core/src/main/java/org/apache/accumulo/fate/Fate.java

```diff
@@ -71,35 +83,58 @@ public void run() {
         try {
           tid = store.reserve();
           TStatus status = store.getStatus(tid);
+          if (status == FAILED) {
+            runnerLog.info(
+                "FATE txid " + Long.toHexString(tid) + " likely cancelled before it started.");
+            return;
+          }
           Repo<T> op = store.top(tid);
-          if (status == TStatus.FAILED_IN_PROGRESS) {
+          if (status == FAILED_IN_PROGRESS) {
             processFailed(tid, op);
           } else {
             Repo<T> prevOp = null;
             try {
+              synchronized (RUNNING_TRANSACTIONS) {
+                RUNNING_TRANSACTIONS.put(tid, Thread.currentThread());
+              }
               deferTime = op.isReady(tid, environment);
 
               if (deferTime == 0) {
                 prevOp = op;
-                if (status == TStatus.SUBMITTED) {
-                  store.setStatus(tid, TStatus.IN_PROGRESS);
+                if (status == SUBMITTED) {
+                  store.setStatus(tid, IN_PROGRESS);
                 }
                 op = op.call(tid, environment);
               } else
                 continue;
 
             } catch (Exception e) {
+              if (e instanceof InterruptedException) {
+                runnerLog.info(
+                    "FATE Runner thread interrupted processing txid " + Long.toHexString(tid));
+                if (prevOp != null) {
+                  processFailed(tid, prevOp);
+                }
+                if (op != prevOp) {
+                  processFailed(tid, op);
+                }
```

Comment:
This code calls processFailed before transitioning the fate tx to FAILED_IN_PROGRESS.  If the manager process dies in the middle of processFailed() then when it comes back it will start executing the tx in the IN_PROGRESS state which could lead to normal processing, failure processing, followed by normal processing for a FATE tx.   This could lead to strange mutations to ZK and metadata table. 

What is the motivation for this special case?

Reply:
Ok, so you are thinking I should change `processFailed` to `transitionToFailed` on lines 116 and 119?

Reply:
I was thinking of just not calling processFailed here unless there is a compelling reason to do so.  When the code calls transitionToFailed then it will eventually call processFailed.

Reply:
Ok, I see. Yeah, I can remove those.

Reply:
Removed in [6d22ea7](https://github.com/apache/accumulo/pull/2467/commits/6d22ea7708b786a9b9cd2aa4469e332e29e57476)